cmd/lotus-shed/indexes.go

```diff
@@ -483,43 +498,186 @@ var inspectEventsCmd = &cli.Command{
+						continue
+					}
+					events, err := api.ChainGetEvents(ctx, *receipt.EventsRoot)
+					if err != nil {
+						return xerrors.Errorf("failed to load events for tipset %s: %w", currTs, err)
+					}
+					if len(events) > 0 {
+						hasEvents = true
+					}
+					expectEvents += len(events)
+					for _, event := range events {
+						expectEntries += len(event.Entries)
+					}
 				}
-				events, err := api.ChainGetEvents(ctx, *receipt.EventsRoot)
-				if err != nil {
-					return fmt.Errorf("failed to load events for tipset %s: %w", currTs, err)
+
+				var actualEvents int
+				if err := stmtEventCount.QueryRow(tsKeyCid.Bytes()).Scan(&actualEvents); err != nil {
+					return xerrors.Errorf("failed to count events for epoch %d (tsk CID %s): %w", ts.Height(), tsKeyCid, err)
 				}
-				expectEvents += len(events)
-				for _, event := range events {
-					expectEntries += len(event.Entries)
+				var actualEntries int
+				if err := stmtEntryCount.QueryRow(tsKeyCid.Bytes()).Scan(&actualEntries); err != nil {
+					return xerrors.Errorf("failed to count entries for epoch %d (tsk CID %s): %w", ts.Height(), tsKeyCid, err)
 				}
-			}
 
-			var problems []string
+				if actualEvents != expectEvents {
+					problems = append(problems, fmt.Sprintf("expected %d events, got %d", expectEvents, actualEvents))
+				}
+				if actualEntries != expectEntries {
+					problems = append(problems, fmt.Sprintf("expected %d entries, got %d", expectEntries, actualEntries))
+				}
+			} else { // compareAmt
+				addrIdCache := make(map[address.Address]abi.ActorID)
+				eventIndex := 0
+				for msgIndex, receipt := range receipts {
+					if receipt.EventsRoot == nil {
+						continue
+					}
+
+					amtRoot, problem, err := func() (cid.Cid, string, error) {
+						events := make([]cbg.CBORMarshaler, 0)
+
+						rows, err := stmtSelectEventIdAndEmitter.Query(tsKeyCid.Bytes(), messages[msgIndex].Cid.Bytes())
+						if err != nil {
+							return cid.Undef, "", xerrors.Errorf("failed to query events: %w", err)
+						}
+						defer func() {
+							_ = rows.Close()
+						}()
+
+						for rows.Next() {
+							var eventId int
+							var emitterAddr []byte
+							if err := rows.Scan(&eventId, &emitterAddr); err != nil {
+								return cid.Undef, "", xerrors.Errorf("failed to scan row: %w", err)
+							}
+
+							addr, err := address.NewFromBytes(emitterAddr)
+							if err != nil {
+								return cid.Undef, "", xerrors.Errorf("failed to parse address: %w", err)
+							}
+							var actorId abi.ActorID
+							if id, ok := addrIdCache[addr]; ok {
+								actorId = id
+							} else {
+								if addr.Protocol() != address.ID {
+									idAddr, err := api.StateLookupID(ctx, addr, ts.Key())
+									if err != nil {
+										// TODO: fix this? we should be able to resolve all addresses
+										return cid.Undef, fmt.Sprintf("failed to resolve address (%s), could not compare amt", addr.String()), nil
+									}
```

Comment:
@Stebalien could you have a look at this and suggest if we can do anything about this? We're reconstructing the event AMT from what we've stored, and we store ID addresses or DelegatedAddresses if it is one. I seem to be able to use `StateLookupID` at the tipset to resolve most of those but it regularly gets hung up on particular addresses it can't resolve. I need an `ActorID` for each of these to reconstruct the event AMT.

Reply:
Deferred execution strikes again, never trust a tipset.

 `LookupIDAddress` uses the parent tipset for performance reasons so you're probably trying to resolve IDs for actors that were created in that tipset.

I'd strongly consider modifying `LookupIDAddress` to use the correct state. Really, I'd audit all calls to `sm.parentState` and fix them to call `stmgr.TipSetState`. This used to be a huge perf issue but we'll now try pretty hard to find the correct state-tree root by walking the chain.

Reply:
Note: that won't help you with _uncle_ blocks. For those, you'll end up re-computing (although you will cache that new state-cid in memory).

A possible optimization is to _try_ looking up the ID in the parent tipset before trying to compute the new tipset. That'll cost more in the worst-case scenario, but may be better in most cases.

Reply:
Ahhh of course, and in this case it's easy enough because (a) we're only dealing with the canonical chain and (b) I have the previous tipset (walking backward) handy so I can just feed that in to `LookupIDAddress` instead.

So making that change gets me all ✔️ for my current chain, no address problems.